src/librustc_mir/transform/acs_propagate.rs

```diff
@@ -40,25 +42,49 @@ use pretty;
 
 #[derive(PartialEq, Debug, Eq, Clone)]
 enum Either<'tcx> {
-    Top,
     Lvalue(Lvalue<'tcx>),
     Const(Constant<'tcx>),
 }
 
-impl<'tcx> Lattice for Either<'tcx> {
-    fn bottom() -> Self { unimplemented!() }
+#[derive(Debug, Clone)]
+enum AcsLattice<'tcx> {
+    Bottom,
```

Comment:
I changed the lattice so that `HashMap::new()` is actually the top element of the lattice - I treat empty elements as top instead of bottom. You'll notice that `join` now does intersection instead of union. This is primarily because when we encounter an unknown function call, we need to mark everything as `Top`, because the function call could change anything. This situation can be improved with alias analysis and information about what the function might do, but the default needs to be `Top`. Additionally, when we enter the function, everything needs to be `Top`. Consider the following case:

```rust
fn foo(mut x: u32) {
    if /* random unoptimizable condition */ {
        x = 5;
    }
    println!("{}", x);
}
```

Previously, since the value stored for `x` was bottom, we would merge the fact `bottom` and the fact `x = 5` at the end of the if statement, concluding that `x` was always `5`. We would optimize to:

```rust
fn foo(mut x: u32) {
    if /* random unoptimizable condition */ {
        x = 5;
    }
    println!("{}", 5);
}
```

which is just wrong.

Note that, in terms of inspiration, while the original paper used a union-based lattice for constant propogation, GHC actually uses an intersection based lattice (see [here](https://ghc.haskell.org/trac/ghc/wiki/Hoopl/Cleanup)).